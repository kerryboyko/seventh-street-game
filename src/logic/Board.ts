import { invert } from "radash";
import Deck from "./poker/Deck";
import Card from "./poker/Card";

// The parks will be at index 0,
// 1 - 7 will be the corresponding streets.

interface Property {
  name: string;
  card: null | Card;
  revealed: boolean;
  owner: string | null;
  scored: boolean;
  price: number;
  type: "ACTION" | "PROPERTY";
  visibleToThesePlayers: Set<string>;
}

interface OutputProperty
  extends Pick<Property, "name" | "revealed" | "owner" | "scored" | "price"> {
  cardDisplay: string;
}

const AVENUES = ["N", "E", "S", "W"];
const generateNeighborhood = (deck: Deck): Record<string, Property> => {
  deck.shuffle();
  const neighborhood: Record<string, Property> = {};
  for (let avenue of AVENUES) {
    for (let i = 0; i < 8; i++) {
      const name = `${avenue}${i}`;
      neighborhood[name] = {
        name,
        card: i === 0 ? null : deck.dealCard(),
        revealed: false,
        visibleToThesePlayers: new Set<string>([]),
        owner: null,
        scored: false,
        price: i * 100,
        type: i === 0 ? "ACTION" : "PROPERTY",
      };
    }
  }
  return neighborhood;
};
export default class Board {
  private deck: Deck;
  private neighborhood;
  public players: string[] = [];
  constructor(public seed?: string) {
    this.deck = new Deck(this.seed);
    this.neighborhood = generateNeighborhood(this.deck);
  }
  public debug = () => ({
    deck: this.deck,
    neighborhood: this.neighborhood,
  });
  public addPlayer = (playerName: string) => {
    if (this.players.includes(playerName)) {
      return false;
    }
    this.players.push(playerName);
    return true;
  };
  public getBoardState = (player?: string): Record<string, OutputProperty> => {
    const output: Record<string, OutputProperty> = {};
    for (const name in this.neighborhood) {
      let {
        card,
        revealed,
        visibleToThesePlayers,
        owner,
        scored,
        price,
        type,
      } = this.neighborhood[name];
      if (type === "ACTION" || card === null) {
        continue;
      }
      const visible = revealed || (player && visibleToThesePlayers.has(player));
      output[name] = {
        name,
        cardDisplay: visible ? card?.display : "??",
        price,
        scored,
        owner,
        revealed,
      };
    }
    return output;
  };
  public setOwner = (
    owner: string,
    propertyName: keyof Board["neighborhood"]
  ) => {
    if (!this.players.includes(owner)) {
      throw new Error(`No player named ${owner} is playing`);
    }
    if (!this.neighborhood[propertyName]) {
      throw new Error(`No property named ${propertyName} exists`);
    }
    this.neighborhood[propertyName].owner = owner;
    this.neighborhood[propertyName].visibleToThesePlayers.add(owner);
  };
}
