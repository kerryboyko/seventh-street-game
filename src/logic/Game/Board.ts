import Deck from "../poker/Deck";
import Card from "../poker/Card";
import Property from "./Property";
// The parks will be at index 0,
// 1 - 7 will be the corresponding streets.

const AVENUES = ["N", "E", "S", "W"];
const generateNeighborhood = (deck: Deck): Record<string, Property> => {
  deck.shuffle();
  const neighborhood: Record<string, Property> = {};
  for (let avenue of AVENUES) {
    for (let i = 0; i < 8; i++) {
      const name = `${avenue}${i}`;
      neighborhood[name] = new Property(
        name,
        i === 0 ? null : deck.dealCard(),
        i * 100,
        i === 0 ? "ACTION" : "PROPERTY"
      );
    }
  }
  return neighborhood;
};
export default class Board {
  private deck: Deck;
  private neighborhood: Record<string, Property>;
  public players: string[] = [];
  constructor(public seed?: string) {
    this.deck = new Deck(this.seed);
    this.neighborhood = generateNeighborhood(this.deck);
  }
  public debug = () => ({
    deck: this.deck,
    neighborhood: this.neighborhood,
  });
  public addPlayer = (playerName: string) => {
    if (this.players.includes(playerName)) {
      return false;
    }
    this.players.push(playerName);
    return true;
  };
  public getBoardState = (player?: string): Record<string, any> => {
    const acc: Record<string, any> = {};
    Object.values(this.neighborhood)
      .filter(({ type }) => type === "PROPERTY")
      .forEach((prop: Property) => {
        acc[prop.name] = prop.getState(player);
      });
      return acc;
  };

  public setOwner = (
    owner: string,
    propertyName: keyof Board["neighborhood"]
  ) => {
    if (!this.players.includes(owner)) {
      throw new Error(`No player named ${owner} is playing`);
    }
    if (!this.neighborhood[propertyName]) {
      throw new Error(`No property named ${propertyName} exists`);
    }
    this.neighborhood[propertyName].setOwner(owner);
  };
}
