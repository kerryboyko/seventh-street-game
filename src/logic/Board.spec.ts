import Board from "./Board";

describe("class Board", () => {
  const board = new Board("seventhst test");
  it("generates a neighborhood", () => {
    expect(board.debug()).toMatchInlineSnapshot(`
      {
        "deck": Deck {
          "cards": [
            Card {
              "rank": 10,
              "suit": "c",
            },
            Card {
              "rank": 11,
              "suit": "d",
            },
            Card {
              "rank": 7,
              "suit": "c",
            },
            Card {
              "rank": 9,
              "suit": "h",
            },
            Card {
              "rank": 3,
              "suit": "s",
            },
            Card {
              "rank": 6,
              "suit": "s",
            },
            Card {
              "rank": 14,
              "suit": "s",
            },
            Card {
              "rank": 5,
              "suit": "s",
            },
            Card {
              "rank": 13,
              "suit": "d",
            },
            Card {
              "rank": 9,
              "suit": "d",
            },
            Card {
              "rank": 5,
              "suit": "c",
            },
            Card {
              "rank": 13,
              "suit": "h",
            },
            Card {
              "rank": 7,
              "suit": "d",
            },
            Card {
              "rank": 4,
              "suit": "d",
            },
            Card {
              "rank": 10,
              "suit": "h",
            },
            Card {
              "rank": 14,
              "suit": "h",
            },
            Card {
              "rank": 8,
              "suit": "h",
            },
            Card {
              "rank": 6,
              "suit": "d",
            },
            Card {
              "rank": 11,
              "suit": "s",
            },
            Card {
              "rank": 13,
              "suit": "s",
            },
            Card {
              "rank": 14,
              "suit": "c",
            },
            Card {
              "rank": 4,
              "suit": "c",
            },
            Card {
              "rank": 2,
              "suit": "d",
            },
            Card {
              "rank": 3,
              "suit": "c",
            },
            Card {
              "rank": 13,
              "suit": "c",
            },
            Card {
              "rank": 8,
              "suit": "d",
            },
            Card {
              "rank": 2,
              "suit": "h",
            },
            Card {
              "rank": 10,
              "suit": "d",
            },
            Card {
              "rank": 12,
              "suit": "h",
            },
            Card {
              "rank": 8,
              "suit": "s",
            },
            Card {
              "rank": 12,
              "suit": "c",
            },
            Card {
              "rank": 4,
              "suit": "s",
            },
            Card {
              "rank": 3,
              "suit": "h",
            },
            Card {
              "rank": 12,
              "suit": "d",
            },
            Card {
              "rank": 3,
              "suit": "d",
            },
            Card {
              "rank": 2,
              "suit": "c",
            },
            Card {
              "rank": 5,
              "suit": "h",
            },
            Card {
              "rank": 9,
              "suit": "c",
            },
            Card {
              "rank": 8,
              "suit": "c",
            },
            Card {
              "rank": 6,
              "suit": "c",
            },
            Card {
              "rank": 10,
              "suit": "s",
            },
            Card {
              "rank": 14,
              "suit": "d",
            },
            Card {
              "rank": 7,
              "suit": "s",
            },
            Card {
              "rank": 2,
              "suit": "s",
            },
            Card {
              "rank": 12,
              "suit": "s",
            },
            Card {
              "rank": 11,
              "suit": "h",
            },
            Card {
              "rank": 11,
              "suit": "c",
            },
            Card {
              "rank": 5,
              "suit": "d",
            },
            Card {
              "rank": 6,
              "suit": "h",
            },
            Card {
              "rank": 4,
              "suit": "h",
            },
            Card {
              "rank": 9,
              "suit": "s",
            },
            Card {
              "rank": 7,
              "suit": "h",
            },
          ],
          "cursor": 32,
          "dealCard": [Function],
          "dealCards": [Function],
          "length": 52,
          "rng": [Function],
          "seed": "seventhst test",
        },
        "neighborhood": {
          "ids": [
            "N0",
            "N1",
            "N2",
            "N3",
            "N4",
            "N5",
            "N6",
            "N7",
            "E0",
            "E1",
            "E2",
            "E3",
            "E4",
            "E5",
            "E6",
            "E7",
            "S0",
            "S1",
            "S2",
            "S3",
            "S4",
            "S5",
            "S6",
            "S7",
            "W0",
            "W1",
            "W2",
            "W3",
            "W4",
            "W5",
            "W6",
            "W7",
          ],
          "neighborhood": {
            "E0": {
              "card": Card {
                "rank": 13,
                "suit": "d",
              },
              "name": "E0",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E1": {
              "card": Card {
                "rank": 9,
                "suit": "d",
              },
              "name": "E1",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E2": {
              "card": Card {
                "rank": 5,
                "suit": "c",
              },
              "name": "E2",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E3": {
              "card": Card {
                "rank": 13,
                "suit": "h",
              },
              "name": "E3",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E4": {
              "card": Card {
                "rank": 7,
                "suit": "d",
              },
              "name": "E4",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E5": {
              "card": Card {
                "rank": 4,
                "suit": "d",
              },
              "name": "E5",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E6": {
              "card": Card {
                "rank": 10,
                "suit": "h",
              },
              "name": "E6",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "E7": {
              "card": Card {
                "rank": 14,
                "suit": "h",
              },
              "name": "E7",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N0": {
              "card": Card {
                "rank": 10,
                "suit": "c",
              },
              "name": "N0",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N1": {
              "card": Card {
                "rank": 11,
                "suit": "d",
              },
              "name": "N1",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N2": {
              "card": Card {
                "rank": 7,
                "suit": "c",
              },
              "name": "N2",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N3": {
              "card": Card {
                "rank": 9,
                "suit": "h",
              },
              "name": "N3",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N4": {
              "card": Card {
                "rank": 3,
                "suit": "s",
              },
              "name": "N4",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N5": {
              "card": Card {
                "rank": 6,
                "suit": "s",
              },
              "name": "N5",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N6": {
              "card": Card {
                "rank": 14,
                "suit": "s",
              },
              "name": "N6",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "N7": {
              "card": Card {
                "rank": 5,
                "suit": "s",
              },
              "name": "N7",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S0": {
              "card": Card {
                "rank": 8,
                "suit": "h",
              },
              "name": "S0",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S1": {
              "card": Card {
                "rank": 6,
                "suit": "d",
              },
              "name": "S1",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S2": {
              "card": Card {
                "rank": 11,
                "suit": "s",
              },
              "name": "S2",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S3": {
              "card": Card {
                "rank": 13,
                "suit": "s",
              },
              "name": "S3",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S4": {
              "card": Card {
                "rank": 14,
                "suit": "c",
              },
              "name": "S4",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S5": {
              "card": Card {
                "rank": 4,
                "suit": "c",
              },
              "name": "S5",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S6": {
              "card": Card {
                "rank": 2,
                "suit": "d",
              },
              "name": "S6",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "S7": {
              "card": Card {
                "rank": 3,
                "suit": "c",
              },
              "name": "S7",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W0": {
              "card": Card {
                "rank": 13,
                "suit": "c",
              },
              "name": "W0",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W1": {
              "card": Card {
                "rank": 8,
                "suit": "d",
              },
              "name": "W1",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W2": {
              "card": Card {
                "rank": 2,
                "suit": "h",
              },
              "name": "W2",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W3": {
              "card": Card {
                "rank": 10,
                "suit": "d",
              },
              "name": "W3",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W4": {
              "card": Card {
                "rank": 12,
                "suit": "h",
              },
              "name": "W4",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W5": {
              "card": Card {
                "rank": 8,
                "suit": "s",
              },
              "name": "W5",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W6": {
              "card": Card {
                "rank": 12,
                "suit": "c",
              },
              "name": "W6",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
            "W7": {
              "card": Card {
                "rank": 4,
                "suit": "s",
              },
              "name": "W7",
              "owner": null,
              "revealed": false,
              "scored": false,
              "visibleToThesePlayers": [],
            },
          },
        },
      }
    `);
  });
});
