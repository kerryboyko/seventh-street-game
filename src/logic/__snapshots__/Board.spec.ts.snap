// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Board Board.getBoardState() can get a board state 1`] = `
{
  "E1": {
    "cardDisplay": "??",
    "name": "E1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
  },
  "E2": {
    "cardDisplay": "??",
    "name": "E2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
  },
  "E3": {
    "cardDisplay": "??",
    "name": "E3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
  },
  "E4": {
    "cardDisplay": "??",
    "name": "E4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
  },
  "E5": {
    "cardDisplay": "??",
    "name": "E5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
  },
  "E6": {
    "cardDisplay": "??",
    "name": "E6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
  },
  "E7": {
    "cardDisplay": "??",
    "name": "E7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
  },
  "N1": {
    "cardDisplay": "??",
    "name": "N1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
  },
  "N2": {
    "cardDisplay": "??",
    "name": "N2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
  },
  "N3": {
    "cardDisplay": "??",
    "name": "N3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
  },
  "N4": {
    "cardDisplay": "??",
    "name": "N4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
  },
  "N5": {
    "cardDisplay": "??",
    "name": "N5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
  },
  "N6": {
    "cardDisplay": "??",
    "name": "N6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
  },
  "N7": {
    "cardDisplay": "??",
    "name": "N7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
  },
  "S1": {
    "cardDisplay": "??",
    "name": "S1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
  },
  "S2": {
    "cardDisplay": "??",
    "name": "S2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
  },
  "S3": {
    "cardDisplay": "??",
    "name": "S3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
  },
  "S4": {
    "cardDisplay": "??",
    "name": "S4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
  },
  "S5": {
    "cardDisplay": "??",
    "name": "S5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
  },
  "S6": {
    "cardDisplay": "??",
    "name": "S6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
  },
  "S7": {
    "cardDisplay": "??",
    "name": "S7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
  },
  "W1": {
    "cardDisplay": "??",
    "name": "W1",
    "owner": null,
    "price": 100,
    "revealed": false,
    "scored": false,
  },
  "W2": {
    "cardDisplay": "??",
    "name": "W2",
    "owner": null,
    "price": 200,
    "revealed": false,
    "scored": false,
  },
  "W3": {
    "cardDisplay": "??",
    "name": "W3",
    "owner": null,
    "price": 300,
    "revealed": false,
    "scored": false,
  },
  "W4": {
    "cardDisplay": "??",
    "name": "W4",
    "owner": null,
    "price": 400,
    "revealed": false,
    "scored": false,
  },
  "W5": {
    "cardDisplay": "??",
    "name": "W5",
    "owner": null,
    "price": 500,
    "revealed": false,
    "scored": false,
  },
  "W6": {
    "cardDisplay": "??",
    "name": "W6",
    "owner": null,
    "price": 600,
    "revealed": false,
    "scored": false,
  },
  "W7": {
    "cardDisplay": "??",
    "name": "W7",
    "owner": null,
    "price": 700,
    "revealed": false,
    "scored": false,
  },
}
`;

exports[`class Board constructor() generates a neighborhood 1`] = `
{
  "deck": Deck {
    "cards": [
      Card {
        "rank": 10,
        "suit": "c",
      },
      Card {
        "rank": 11,
        "suit": "d",
      },
      Card {
        "rank": 7,
        "suit": "c",
      },
      Card {
        "rank": 9,
        "suit": "h",
      },
      Card {
        "rank": 3,
        "suit": "s",
      },
      Card {
        "rank": 6,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "s",
      },
      Card {
        "rank": 5,
        "suit": "s",
      },
      Card {
        "rank": 13,
        "suit": "d",
      },
      Card {
        "rank": 9,
        "suit": "d",
      },
      Card {
        "rank": 5,
        "suit": "c",
      },
      Card {
        "rank": 13,
        "suit": "h",
      },
      Card {
        "rank": 7,
        "suit": "d",
      },
      Card {
        "rank": 4,
        "suit": "d",
      },
      Card {
        "rank": 10,
        "suit": "h",
      },
      Card {
        "rank": 14,
        "suit": "h",
      },
      Card {
        "rank": 8,
        "suit": "h",
      },
      Card {
        "rank": 6,
        "suit": "d",
      },
      Card {
        "rank": 11,
        "suit": "s",
      },
      Card {
        "rank": 13,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "c",
      },
      Card {
        "rank": 4,
        "suit": "c",
      },
      Card {
        "rank": 2,
        "suit": "d",
      },
      Card {
        "rank": 3,
        "suit": "c",
      },
      Card {
        "rank": 13,
        "suit": "c",
      },
      Card {
        "rank": 8,
        "suit": "d",
      },
      Card {
        "rank": 2,
        "suit": "h",
      },
      Card {
        "rank": 10,
        "suit": "d",
      },
      Card {
        "rank": 12,
        "suit": "h",
      },
      Card {
        "rank": 8,
        "suit": "s",
      },
      Card {
        "rank": 12,
        "suit": "c",
      },
      Card {
        "rank": 4,
        "suit": "s",
      },
      Card {
        "rank": 3,
        "suit": "h",
      },
      Card {
        "rank": 12,
        "suit": "d",
      },
      Card {
        "rank": 3,
        "suit": "d",
      },
      Card {
        "rank": 2,
        "suit": "c",
      },
      Card {
        "rank": 5,
        "suit": "h",
      },
      Card {
        "rank": 9,
        "suit": "c",
      },
      Card {
        "rank": 8,
        "suit": "c",
      },
      Card {
        "rank": 6,
        "suit": "c",
      },
      Card {
        "rank": 10,
        "suit": "s",
      },
      Card {
        "rank": 14,
        "suit": "d",
      },
      Card {
        "rank": 7,
        "suit": "s",
      },
      Card {
        "rank": 2,
        "suit": "s",
      },
      Card {
        "rank": 12,
        "suit": "s",
      },
      Card {
        "rank": 11,
        "suit": "h",
      },
      Card {
        "rank": 11,
        "suit": "c",
      },
      Card {
        "rank": 5,
        "suit": "d",
      },
      Card {
        "rank": 6,
        "suit": "h",
      },
      Card {
        "rank": 4,
        "suit": "h",
      },
      Card {
        "rank": 9,
        "suit": "s",
      },
      Card {
        "rank": 7,
        "suit": "h",
      },
    ],
    "cursor": 28,
    "dealCard": [Function],
    "dealCards": [Function],
    "length": 52,
    "rng": [Function],
    "seed": "seventhst test",
  },
  "neighborhood": {
    "E0": {
      "card": null,
      "name": "E0",
      "owner": null,
      "price": 0,
      "revealed": false,
      "scored": false,
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "E1": {
      "card": Card {
        "rank": 5,
        "suit": "s",
      },
      "name": "E1",
      "owner": null,
      "price": 100,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E2": {
      "card": Card {
        "rank": 13,
        "suit": "d",
      },
      "name": "E2",
      "owner": null,
      "price": 200,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E3": {
      "card": Card {
        "rank": 9,
        "suit": "d",
      },
      "name": "E3",
      "owner": null,
      "price": 300,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E4": {
      "card": Card {
        "rank": 5,
        "suit": "c",
      },
      "name": "E4",
      "owner": null,
      "price": 400,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E5": {
      "card": Card {
        "rank": 13,
        "suit": "h",
      },
      "name": "E5",
      "owner": null,
      "price": 500,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E6": {
      "card": Card {
        "rank": 7,
        "suit": "d",
      },
      "name": "E6",
      "owner": null,
      "price": 600,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "E7": {
      "card": Card {
        "rank": 4,
        "suit": "d",
      },
      "name": "E7",
      "owner": null,
      "price": 700,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N0": {
      "card": null,
      "name": "N0",
      "owner": null,
      "price": 0,
      "revealed": false,
      "scored": false,
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "N1": {
      "card": Card {
        "rank": 10,
        "suit": "c",
      },
      "name": "N1",
      "owner": null,
      "price": 100,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N2": {
      "card": Card {
        "rank": 11,
        "suit": "d",
      },
      "name": "N2",
      "owner": null,
      "price": 200,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N3": {
      "card": Card {
        "rank": 7,
        "suit": "c",
      },
      "name": "N3",
      "owner": null,
      "price": 300,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N4": {
      "card": Card {
        "rank": 9,
        "suit": "h",
      },
      "name": "N4",
      "owner": null,
      "price": 400,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N5": {
      "card": Card {
        "rank": 3,
        "suit": "s",
      },
      "name": "N5",
      "owner": null,
      "price": 500,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N6": {
      "card": Card {
        "rank": 6,
        "suit": "s",
      },
      "name": "N6",
      "owner": null,
      "price": 600,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "N7": {
      "card": Card {
        "rank": 14,
        "suit": "s",
      },
      "name": "N7",
      "owner": null,
      "price": 700,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S0": {
      "card": null,
      "name": "S0",
      "owner": null,
      "price": 0,
      "revealed": false,
      "scored": false,
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "S1": {
      "card": Card {
        "rank": 10,
        "suit": "h",
      },
      "name": "S1",
      "owner": null,
      "price": 100,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S2": {
      "card": Card {
        "rank": 14,
        "suit": "h",
      },
      "name": "S2",
      "owner": null,
      "price": 200,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S3": {
      "card": Card {
        "rank": 8,
        "suit": "h",
      },
      "name": "S3",
      "owner": null,
      "price": 300,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S4": {
      "card": Card {
        "rank": 6,
        "suit": "d",
      },
      "name": "S4",
      "owner": null,
      "price": 400,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S5": {
      "card": Card {
        "rank": 11,
        "suit": "s",
      },
      "name": "S5",
      "owner": null,
      "price": 500,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S6": {
      "card": Card {
        "rank": 13,
        "suit": "s",
      },
      "name": "S6",
      "owner": null,
      "price": 600,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "S7": {
      "card": Card {
        "rank": 14,
        "suit": "c",
      },
      "name": "S7",
      "owner": null,
      "price": 700,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W0": {
      "card": null,
      "name": "W0",
      "owner": null,
      "price": 0,
      "revealed": false,
      "scored": false,
      "type": "ACTION",
      "visibleToThesePlayers": Set {},
    },
    "W1": {
      "card": Card {
        "rank": 4,
        "suit": "c",
      },
      "name": "W1",
      "owner": null,
      "price": 100,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W2": {
      "card": Card {
        "rank": 2,
        "suit": "d",
      },
      "name": "W2",
      "owner": null,
      "price": 200,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W3": {
      "card": Card {
        "rank": 3,
        "suit": "c",
      },
      "name": "W3",
      "owner": null,
      "price": 300,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W4": {
      "card": Card {
        "rank": 13,
        "suit": "c",
      },
      "name": "W4",
      "owner": null,
      "price": 400,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W5": {
      "card": Card {
        "rank": 8,
        "suit": "d",
      },
      "name": "W5",
      "owner": null,
      "price": 500,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W6": {
      "card": Card {
        "rank": 2,
        "suit": "h",
      },
      "name": "W6",
      "owner": null,
      "price": 600,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
    "W7": {
      "card": Card {
        "rank": 10,
        "suit": "d",
      },
      "name": "W7",
      "owner": null,
      "price": 700,
      "revealed": false,
      "scored": false,
      "type": "PROPERTY",
      "visibleToThesePlayers": Set {},
    },
  },
}
`;
